src/handlers/assign.rs

```diff
@@ -708,14 +781,57 @@ fn find_reviewer_from_names(
     //
     // These are all ideas for improving the selection here. However, I'm not
     // sure they are really worth the effort.
-    Ok(candidates
+
+    // filter out team members without capacity
+    let filtered_candidates = filter_by_capacity(db, &candidates)
+        .await
+        .expect("Error while filtering out team members");
+
+    if filtered_candidates.is_empty() {
+        return Err(FindReviewerError::AllReviewersFiltered {
+            initial: names.to_vec(),
+            filtered: names.to_vec(),
+        });
+    }
+
+    log::debug!("Filtered list of candidates: {:?}", filtered_candidates);
+
+    Ok(filtered_candidates
         .into_iter()
         .choose(&mut rand::thread_rng())
-        .expect("candidate_reviewers_from_names always returns at least one entry")
+        .expect("candidate_reviewers_from_names should return at least one entry")
         .to_string())
 }
 
-/// Returns a list of candidate usernames to choose as a reviewer.
+/// Filter out candidates not having review capacity
+async fn filter_by_capacity(
+    db: &DbClient,
+    candidates: &HashSet<&str>,
+) -> anyhow::Result<HashSet<String>> {
+    let usernames = candidates
+        .iter()
+        .map(|c| *c)
+        .collect::<Vec<&str>>()
+        .join(",");
+
+    let q = format!(
+        "
+SELECT username
+FROM review_prefs r
+JOIN users on users.user_id=r.user_id
+AND username = ANY('{{ {} }}')
+AND CARDINALITY(r.assigned_prs) < LEAST(COALESCE(max_assigned_prs,1000000))",
```

Comment:
A note about this SQL dance, it is a safeguard against *null* fields to be able to do a number comparison. [CARDINALITY](https://www.postgresql.org/docs/16/arrays.html#ARRAYS-ACCESSING) returns the number or array items even when empty, `LEAST(COALESCE(...))` returns a number when `max_assigned_prs` is *null*.

Reply:
Does this need to be `r.max_assigned_prs`? (If not, should it for clarity?)

Reply:
Needs not since it's not ambigous but good for clarity :+1: